File: src/pizza_interface.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "../include/pizza_interface.h"
#include "../include/pizza_types.h"

#define MENU_FILE_PATH "data/pizza_menu.txt"

void display_main_menu(void) {
    printf("\n=== Pizza Management System ===\n");
    printf("1. Manage Pizza Menu\n");
    printf("2. Create Order & Calculate Split\n");
    printf("3. Exit\n");
    printf("Choose an option (1-3): ");
}

void manage_pizza_menu(PizzaMenu *menu) {
    int choice;
    char name[MAX_PIZZA_NAME_LENGTH];
    double price;
    
    do {
        printf("\n=== Pizza Menu Management ===\n");
        printf("1. View Pizza Menu\n");
        printf("2. Add New Pizza Type\n");
        printf("3. Remove Pizza Type\n");
        printf("4. Back to Main Menu\n");
        printf("Choose an option (1-4): ");
        
        choice = get_user_choice(1, 4);
        
        switch (choice) {
            case 1:
                display_pizza_menu(menu);
                break;
                
            case 2:
                printf("Enter pizza name: ");
                clear_input_buffer();
                if (fgets(name, sizeof(name), stdin) == NULL) {
                    printf("Error: Failed to read input.\n");
                    break;
                }
                name[strcspn(name, "\n")] = '\0'; // Remove newline
                
                // Validate pizza name length and content
                if (strlen(name) == 0) {
                    printf("Error: Pizza name cannot be empty.\n");
                    break;
                }
                
                printf("Enter price (€): ");
                if (scanf("%lf", &price) != 1 || price < MIN_PRICE || price > MAX_PRICE) {
                    printf("Error: Invalid price (must be €%.2f-€%.2f).\n", MIN_PRICE, MAX_PRICE);
                    clear_input_buffer();
                    break;
                }
                
                add_pizza_type(menu, name, price);
                if (!save_pizza_menu(menu, MENU_FILE_PATH)) {
                    printf("Warning: Could not save menu to file.\n");
                }
                break;
                
            case 3:
                display_pizza_menu(menu);
                printf("Enter pizza name to remove: ");
                clear_input_buffer();
                if (fgets(name, sizeof(name), stdin) == NULL) {
                    printf("Error: Failed to read input.\n");
                    break;
                }
                name[strcspn(name, "\n")] = '\0'; // Remove newline
                
                // Validate input
                if (strlen(name) == 0) {
                    printf("Error: Pizza name cannot be empty.\n");
                    break;
                }
                
                remove_pizza_type(menu, name);
                if (!save_pizza_menu(menu, MENU_FILE_PATH)) {
                    printf("Warning: Could not save menu to file.\n");
                }
                break;
                
            case 4:
                printf("Returning to main menu...\n");
                break;
        }
    } while (choice != 4);
}

void create_order(const PizzaMenu *menu, PizzaOrder *order) {
    order->item_count = 0;
    order->total_amount = 0.0;
    
    printf("\n=== Create Pizza Order ===\n");
    display_pizza_menu(menu);
    
    int continue_ordering = 1;
    while (continue_ordering && order->item_count < MAX_PIZZA_TYPES) {
        printf("Enter pizza ID (1-%d) or 0 to finish: ", menu->count);
        int pizza_id = get_user_choice(0, menu->count);
        
        if (pizza_id == 0) {
            continue_ordering = 0;
            break;
        }
        
        int pizza_index = pizza_id - 1;
        if (!menu->pizzas[pizza_index].available) {
            printf("Sorry, %s is not available.\n", menu->pizzas[pizza_index].name);
            continue;
        }
        
        printf("Enter quantity for %s: ", menu->pizzas[pizza_index].name);
        int quantity;
        if (scanf("%d", &quantity) != 1 || quantity <= 0 || quantity > MAX_QUANTITY) {
            printf("Error: Invalid quantity (must be 1-%d).\n", MAX_QUANTITY);
            clear_input_buffer();
            continue;
        }
        
        // Check for potential integer overflow
        double item_total = quantity * menu->pizzas[pizza_index].price;
        if (order->total_amount + item_total < order->total_amount) {
            printf("Error: Order total would overflow. Please reduce quantity.\n");
            continue;
        }
        
        // Add to order using safe string copy
        strncpy(order->items[order->item_count].pizza_name, menu->pizzas[pizza_index].name, MAX_PIZZA_NAME_LENGTH - 1);
        order->items[order->item_count].pizza_name[MAX_PIZZA_NAME_LENGTH - 1] = '\0';
        order->items[order->item_count].quantity = quantity;
        order->items[order->item_count].unit_price = menu->pizzas[pizza_index].price;
        order->total_amount += item_total;
        order->item_count++;
        
        printf("Added %d x %s to order.\n", quantity, menu->pizzas[pizza_index].name);
        
        printf("Continue ordering? (1=Yes, 0=No): ");
        continue_ordering = get_user_choice(0, 1);
    }
    
    if (order->item_count == 0) {
        printf("No items in order.\n");
        return;
    }
    
    printf("Enter number of people splitting the bill: ");
    if (scanf("%d", &order->num_people) != 1 || order->num_people <= 0) {
        printf("Error: Invalid number of people.\n");
        order->num_people = 1;
    }
}

void display_order_summary(const PizzaOrder *order) {
    if (order->item_count == 0) {
        printf("No items in order.\n");
        return;
    }
    
    printf("\n=== Order Summary ===\n");
    printf("%-25s %-8s %-10s %-10s\n", "Pizza", "Qty", "Unit Price", "Subtotal");
    printf("-------------------------------------------------------\n");
    
    for (int i = 0; i < order->item_count; i++) {
        double subtotal = order->items[i].quantity * order->items[i].unit_price;
        printf("%-25s %-8d €%-9.2f €%-9.2f\n", 
               order->items[i].pizza_name,
               order->items[i].quantity,
               order->items[i].unit_price,
               subtotal);
    }
    
    printf("-------------------------------------------------------\n");
    printf("Total Amount: €%.2f\n", order->total_amount);
    printf("Number of People: %d\n", order->num_people);
}

void calculate_bill_split(const PizzaOrder *order) {
    if (order->item_count == 0 || order->num_people <= 0) {
        printf("Invalid order or number of people.\n");
        return;
    }
    
    double cost_per_person = order->total_amount / order->num_people;
    
    printf("\n=== Bill Split Results ===\n");
    printf("Total bill: €%.2f\n", order->total_amount);
    printf("Number of people: %d\n", order->num_people);
    printf("Cost per person: €%.2f\n", cost_per_person);
}

int get_user_choice(int min, int max) {
    int choice;
    while (scanf("%d", &choice) != 1 || choice < min || choice > max) {
        printf("Invalid input. Please enter a number between %d and %d: ", min, max);
        clear_input_buffer();
    }
    return choice;
}

void clear_input_buffer(void) {
    int c;
    while ((c = getchar()) != '\n' && c != EOF);
}

File: python-demo/src/pizza_interface.py
"""
Pizza Interface Module
Handles user interaction with secure input validation
"""

import sys
from decimal import Decimal, InvalidOperation
from .pizza_types import PizzaMenu
from .pizza_order import PizzaOrder

class PizzaInterface:
    """Manages user interface for pizza management system"""
    
    def __init__(self):
        self.menu = PizzaMenu()
        self.menu_file = "data/pizza_menu.json"
        
    def run(self):
        """Main application loop"""
        print("=== Pizza Management System ===")
        print("Loading pizza menu...")
        
        # Load menu from file or use defaults
        if not self.menu.load_from_file(self.menu_file):
            print("Creating default pizza menu...")
            self.menu.save_to_file(self.menu_file)
        
        print(f"Pizza menu loaded successfully! ({self.menu.count()} pizza types available)")
        
        while True:
            try:
                choice = self._display_main_menu()
                
                if choice == 1:
                    self._manage_pizza_menu()
                elif choice == 2:
                    self._create_order()
                elif choice == 3:
                    print("Thank you for using Pizza Management System!")
                    print("Saving menu...")
                    if not self.menu.save_to_file(self.menu_file):
                        print("Warning: Could not save menu to file.")
                    break
                
            except KeyboardInterrupt:
                print("\n\nExiting...")
                break
            except Exception as e:
                print(f"An error occurred: {e}")
    
    def _display_main_menu(self) -> int:
        """Display main menu and get user choice"""
        print("\n=== Pizza Management System ===")
        print("1. Manage Pizza Menu")
        print("2. Create Order & Calculate Split")
        print("3. Exit")
        
        return self._get_user_choice("Choose an option (1-3): ", 1, 3)
    
    def _manage_pizza_menu(self):
        """Handle pizza menu management"""
        while True:
            try:
                print("\n=== Pizza Menu Management ===")
                print("1. View Pizza Menu")
                print("2. Add New Pizza Type")
                print("3. Remove Pizza Type")
                print("4. Back to Main Menu")
                
                choice = self._get_user_choice("Choose an option (1-4): ", 1, 4)
                
                if choice == 1:
                    self.menu.display_menu()
                elif choice == 2:
                    self._add_pizza_type()
                elif choice == 3:
                    self._remove_pizza_type()
                elif choice == 4:
                    break
                    
            except Exception as e:
                print(f"Error: {e}")
    
    def _add_pizza_type(self):
        """Add a new pizza type"""
        try:
            print("Enter pizza name: ", end="")
            name = self._get_safe_input().strip()
            
            if not name:
                print("Error: Pizza name cannot be empty.")
                return
            
            price_str = self._get_safe_input("Enter price (€): ")
            try:
                price = Decimal(price_str)
            except InvalidOperation:
                print("Error: Invalid price format.")
                return
            
            self.menu.add_pizza_type(name, price)
            print(f"Pizza type '{name}' added successfully!")
            
            if not self.menu.save_to_file(self.menu_file):
                print("Warning: Could not save menu to file.")
                
        except ValueError as e:
            print(f"Error: {e}")
        except Exception as e:
            print(f"Unexpected error: {e}")
    
    def _remove_pizza_type(self):
        """Remove a pizza type"""
        try:
            self.menu.display_menu()
            
            if self.menu.count() == 0:
                print("No pizzas to remove.")
                return
            
            name = self._get_safe_input("Enter pizza name to remove: ").strip()
            
            if not name:
                print("Error: Pizza name cannot be empty.")
                return
            
            self.menu.remove_pizza_type(name)
            print(f"Pizza type '{name}' removed successfully!")
            
            if not self.menu.save_to_file(self.menu_file):
                print("Warning: Could not save menu to file.")
                
        except ValueError as e:
            print(f"Error: {e}")
        except Exception as e:
            print(f"Unexpected error: {e}")
    
    def _create_order(self):
        """Create and process an order"""
        try:
            order = PizzaOrder()
            
            print("\n=== Create Pizza Order ===")
            self.menu.display_menu()
            
            if self.menu.count() == 0:
                print("No pizzas available to order.")
                return
            
            # Order items
            while True:
                try:
                    pizza_id = self._get_user_choice(
                        f"Enter pizza ID (1-{self.menu.count()}) or 0 to finish: ",
                        0, self.menu.count()
                    )
                    
                    if pizza_id == 0:
                        break
                    
                    pizza = self.menu.get_pizza_by_index(pizza_id)
                    if not pizza:
                        print("Invalid pizza ID.")
                        continue
                    
                    if not pizza.available:
                        print(f"Sorry, {pizza.name} is not available.")
                        continue
                    
                    quantity = self._get_user_choice(
                        f"Enter quantity for {pizza.name}: ",
                        1, 1000
                    )
                    
                    order.add_item(pizza, quantity)
                    print(f"Added {quantity} x {pizza.name} to order.")
                    
                    continue_ordering = self._get_user_choice(
                        "Continue ordering? (1=Yes, 0=No): ",
                        0, 1
                    )
                    
                    if continue_ordering == 0:
                        break
                        
                except ValueError as e:
                    print(f"Error: {e}")
                    continue
            
            if order.item_count() == 0:
                print("No items in order.")
                return
            
            # Get number of people
            num_people = self._get_user_choice(
                "Enter number of people splitting the bill: ",
                1, 1000
            )
            
            order.set_num_people(num_people)
            
            # Display results
            order.display_summary()
            order.calculate_bill_split()
            
        except Exception as e:
            print(f"Error creating order: {e}")
    
    def _get_user_choice(self, prompt: str, min_val: int, max_val: int) -> int:
        """Get validated integer input from user"""
        while True:
            try:
                choice_str = self._get_safe_input(prompt)
                choice = int(choice_str)
                
                if min_val <= choice <= max_val:
                    return choice
                else:
                    print(f"Invalid input. Please enter a number between {min_val} and {max_val}.")
                    
            except ValueError:
                print(f"Invalid input. Please enter a number between {min_val} and {max_val}.")
            except KeyboardInterrupt:
                raise
            except Exception:
                print("Invalid input. Please try again.")
    
    def _get_safe_input(self, prompt: str = "") -> str:
        """Get safe input from user with length limits"""
        if prompt:
            print(prompt, end="")
        
        try:
            user_input = input()
            
            # Limit input length to prevent memory issues
            if len(user_input) > 1000:
                raise ValueError("Input too long")
            
            return user_input
            
        except EOFError:
            print("\nInput terminated.")
            sys.exit(0)
        except KeyboardInterrupt:
            raise
